src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,550 @@
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
+
+        src_file = _get_source_file(assembly_dir_in_csd, assembly_name)
+
+        assembly_id = os.path.basename(assembly_dir_in_csd)
+        assembly_ids.append(assembly_id)
+        assembly_dir_in_ws = os.path.join(output_dir, assembly_id)
+        os.makedirs(assembly_dir_in_ws, exist_ok=True)
+
+        dest_file = os.path.join(assembly_dir_in_ws, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+
+    return assembly_ids
+
+
+def upload_assemblies_to_workspace(
+        conf: Conf,
+        workspace_name: str,
+        data_dir: str,
+) -> list[str]:
+    """
+    Upload assemblies to workspace and record failed assembly names.
+
+    conf: Conf object
+    workspace_name: a string used as a name for a workspace
+    data_dir: directory of assemblies to upload
+    """
+
+    assembly_files = os.listdir(data_dir)
+    print(f'start uploading {len(assembly_files)} assembly files')
+
+    failed_names = list()
+
+    counter = 1
+    for assembly_name in assembly_files:
+
+        if counter % 5000 == 0:
+            print(f"{round(counter / len(assembly_files), 4) * 100}% finished at {datetime.now()}")
+
+        assembly_path = os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, DATA_DIR, assembly_name)
+        try:
+            _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+        except Exception as e:
+            print(e)
+            failed_names.append(assembly_name)
+
+        counter += 1
+
+    if failed_names:
+        print(f'Failed to upload {failed_names}')
+
+    return failed_names
+
+
+def create_entries_in_ws_and_softlinks_in_csd(
+        csd_upload: str,
+        output_dir: str,
+        assembly_names: list[str],
+        all_assemblies: dict[str, str],
+) -> None:
+    """
+    Create standard entries in workspace and softlinks in collectionssource directory.
+
+    csd_upload: directory of assemblies that have been uploaded to workspace
+    output_dir: output directory to create entries in workspace
+    assembly_names: a list of assembly names that exist in target workspace
+    all_assemblies: a dictionary of assembly name to file path
+    """
+    assembly_ids = create_entries_in_sd_workspace(assembly_names, all_assemblies, output_dir)
+    loader_helper.create_softlinks_in_csd(csd_upload, output_dir, assembly_ids)
+
+
+def main():
+    parser = _get_parser()
+    args = parser.parse_args()
+
+    workspace_id = args.workspace_id
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = args.root_dir
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    keep_job_dir = args.keep_job_dir
+    overwrite = args.overwrite
```

Comment:
I will remove overwrite arg for now.

Reply:
Updated!

Reply:
Before we resolve this I'd still like to get @Tianhao-Gu's thoughts

Reply:
I think throw an error makes more sense here.

Reply:
Updated. An error will be thrown if try to upload an assembly file that's already uploaded.

Reply:
UUIC the way this works is that if the YAML file in the source dir indicates the file is already uploaded it's skipped. If there's a race condition where 2 instances of the uploader are running that means it's possible we could create two versions of the same object, and then one of the instances would fail when it tries to update the YAML file since the other one got there first (or if they're both right at the same time it's theoretically possible their reads could clobber each other.

@Tianhao-Gu do you think this is ok?

Reply:
yea. I think that works.

Reply:
:+1: